include/tsar/Analysis/Memory/MemoryLocationRange.h

```diff
@@ -125,9 +165,352 @@ struct MemoryLocationRange {
+      }
+    }
+
+    Node *mSource;
+    llvm::SmallVector<llvm::SmallVector<Node, 3>, 4> mNodes;
+    LocationList *mLC, *mRC;
+  };
+
+  inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS,
+                                       const DimensionGraph &G) {
+    OS << "Dimension Graph:\n";
+    std::size_t N = 0;
+    for (auto &Level : G.mNodes) {
+      OS << "Level " << N << ":\n";
+      for (auto &Node : Level) {
+        auto &Dim = Node.Dim;
+        OS << "[" << &Node << "] ";
+        OS << "(" << (Node.IsIntersection ? "I" : "N") << ")";
+        OS << "{Start: " << Dim.Start << ", Step: " << Dim.Step <<
+            ", TripCount: " << Dim.TripCount << ", DimSize: " <<
+            Dim.DimSize << "} ";
+        OS << "[L: ";
+        for (auto *Child : Node.LeftChildren)
+          OS << Child << ", ";
+        OS << "][R: ";
+        for (auto *Child : Node.RightChildren)
+          OS << Child << ", ";
+        OS << "]\n";
+      }
+      ++N;
+      OS << "\n";
+    }
+    OS << "Has Left: " << (G.mLC ? "Yes" : "No") << "\n";
+    OS << "Has Right: " << (G.mRC ? "Yes" : "No") << "\n";
+    return OS;
+  }
+
+  inline void printSolutionInfo(llvm::raw_ostream &OS,
+                                const MemoryLocationRange &Int,
+                                const DimensionGraph &G) {
+    auto PrintRange = [&OS](const MemoryLocationRange &R) {
+      auto &Dim = R.DimList[0];
+      OS << "{" << (R.Ptr == nullptr ? "Empty" : "Full") << " | ";
+      OS << Dim.Start << " + " << Dim.Step << " * T, T in [" << 0 <<
+          ", " << Dim.TripCount << ")} ";
+    };
+    OS << "[EQUATION] Solution:\n" << G;
+    OS << "Left: ";
+    if (G.mLC)
+      for (auto &R : *G.mLC) {
+        PrintRange(R);
+      }
+    OS << "\nIntersection: ";
+    PrintRange(Int);
+    OS << "\nRight: ";
+    if (G.mRC)
+      for (auto &R : *G.mRC) {
+        PrintRange(R);
+      }
+    OS << "\n[EQUATION] Solution has been printed.\n";
+  }
+
+  /// \brief Finds an intersection between non-scalar memory locations LHS and
+  /// RHS.
+  ///
+  /// \param [in] LHS The first location to intersect.
+  /// \param [in] RHS The second location to intersect. 
+  /// \param [out] Int The result of intersection. If intersection is empty, 
+  /// Int.Ptr will be set to `nullptr`.
+  /// \param [out] LC List of memory locations to which the difference 
+  /// between locations LHS and Int will be added. It will not change if the
+  /// intersection is empty.
+  /// \param [out] RC List of memory locations to which the difference 
+  /// between locations RHS and Int will be added. It will not change if the
+  /// intersection is empty.
+  /// \return `true` if intersection is not empty, `false` otherwise.
+  inline bool intersect(const MemoryLocationRange &LHS,
+      const MemoryLocationRange &RHS, MemoryLocationRange &Int,
+      llvm::SmallVector<MemoryLocationRange, 0> *LC,
+      llvm::SmallVector<MemoryLocationRange, 0> *RC) {
```

Comment:
Стандартный способ передать `SmallVector` произвольного размера:

```suggestion
      llvm::SmallVectorImpl<MemoryLocationRange> *LC,
      llvm::SmallVectorImpl<MemoryLocationRange> *RC) {
```

`LC` и `RC` можно не указывать, т.е. можно указать `nullptr`, чтобы они не были посчитаны? Если так, то стоит это в описание параметров добавить, в комментарии.